Sources/TuistLoader/Mappers/ProjectDescriptionHelpersLocationMapper.swift

```diff
@@ -0,0 +1,127 @@
+        guard let helpersLocation = config.helpersLocation else {
+            return (project, [])
+        }
+
+        logger.notice("Using \(helpersLocation) for ProjectDescriptionHelpers")
+
+        switch helpersLocation {
+        case let .local(path):
+            return (project, try collectLocalHelpers(at: path))
+        case let .gitWithBranch(url, gitId),
+             let .gitWithHash(url, gitId),
+             let .gitWithTag(url, gitId):
+            return (project, try collectGitHelpers(at: url, gitId: gitId))
+        }
+    }
+
+    /// Collects files from a local path and copies them to the `ProjectDescriptionHelpers` location.
+    private func collectLocalHelpers(at path: String) throws -> [SideEffectDescriptor] {
+        guard let derivedHelpersDirectory = derivedHelpersDirectory else {
+            logger.error("Unable to locate derived helpers directory.")
+            return []
+        }
+
+        let localHelpersPath = AbsolutePath(path)
+
+        let sources = try FileHandler.shared
+            .contentsOfDirectory(localHelpersPath)
+            .filter { $0.extension == "swift" }
+
+        let sourceContents = try sources.map { try FileHandler.shared.readFile($0) }
+        let sourceNames = sources.map(\.basename)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+
+        let createFiles = zip(sourceNames, sourceContents)
+            .map { FileDescriptor(path: AbsolutePath(derivedHelpersDirectory, $0), contents: $1, state: .present) }
+            .map(SideEffectDescriptor.file)
+
+        return [.directory(createHelpersDirectory)] + createFiles
+    }
+
+    private func collectGitHelpers(at url: String, gitId: String) throws -> [SideEffectDescriptor] {
+        guard let derivedHelpersDirectory = derivedHelpersDirectory else {
+            logger.error("Unable to locate derived helpers directory.")
+            return []
+        }
+
+        let helpersGitDirectory = derivedHelpersDirectory.appending(.init(".git"))
+
+        let deleteHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .absent)
+        let createHelpersDirectory = DirectoryDescriptor(path: derivedHelpersDirectory, state: .present)
+        let gitClone = CommandDescriptor(command: ["git", "clone", url, derivedHelpersDirectory.pathString])
+        let gitCheckout = CommandDescriptor(command: [
+            "git",
+            "--git-dir",
+            helpersGitDirectory.pathString,
+            "--work-tree",
+            derivedHelpersDirectory.pathString,
+            "checkout",
+            gitId
+        ])
+        let removeGit = DirectoryDescriptor(path: helpersGitDirectory, state: .absent)
+        
+        return [
+            .directory(deleteHelpersDirectory),
+            .directory(createHelpersDirectory),
+            .command(gitClone),
+            .command(gitCheckout),
+            .directory(removeGit)
+        ]
+    }
```

Comment:
Hey @fortmarek How did you envision using `GitController` here? From my understanding we're supposed to return `SideEffectDescriptors` in the mappers instead of explicitly performing these operations. Let me know if I'm missing something!

Reply:
Nope, this is the correct solution -> I have not realized that the git commands are going to be used in `Mapper` context 🙂